crates/xline/src/storage/kvwatcher.rs

```diff
@@ -136,26 +138,29 @@ impl Watcher {
             revision,
             compacted: self.compacted,
         };
-        if !self.compacted {
-            if revision < self.start_rev || 0 == events_len {
-                return Ok(());
-            }
-            debug!(watch_id, revision, events_len, "try to send watch response");
+        if !self.compacted
+            && (revision < self.start_rev
+                || self.notified_set.contains(&revision)
+                || 0 == events_len)
+        {
+            return Ok(());
         };
 
         match self.event_tx.try_send(watch_event) {
             Ok(_) => {
-                debug!(watch_id, revision, "response sent successfully");
-                self.start_rev = revision.overflow_add(1);
+                assert!(
+                    self.notified_set.insert(revision),
+                    "should not notify the same revision twice!!"
+                );
```

Comment:
The revision is checked in L143, so the assert is not needed here.

Reply:
Done